File: Sigmath/Sigmath/Sigmath.ConvertedToVBNet/imp/topology.vb
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks

Imports System.Collections.Generic
Imports System.Linq

Class Topology
	' Function to check if a space is connected
	Public Shared Function IsConnected(adjacencyList As List(Of Integer)()) As Boolean
		Dim numberOfVertices As Integer = adjacencyList.Length
		Dim visited As Boolean() = New Boolean(numberOfVertices - 1) {}

		' Perform depth-first search starting from the first vertex
		DFS(adjacencyList, 0, visited)

		' Check if all vertices are visited
		Return visited.All(Function(v) v)
	End Function

	' Depth-first search to mark connected vertices
	Private Shared Sub DFS(adjacencyList As List(Of Integer)(), vertex As Integer, visited As Boolean())
		visited(vertex) = True

		For Each neighbor As var In adjacencyList(vertex)
			If Not visited(neighbor) Then
				DFS(adjacencyList, neighbor, visited)
			End If
		Next
	End Sub

	' Function to check if a space is compact
	Public Shared Function IsCompact(openCover As List(Of Integer)()) As Boolean
		For Each [set] As var In openCover
			If Not [set].Any() Then
					' An open cover must be non-empty
				Return False
			End If
		Next

		' Additional conditions for compactness can be added based on the specific context
		Return True
	End Function

	' Function to calculate the Euler characteristic of a space
	Public Shared Function EulerCharacteristic(vertices As Integer, edges As Integer, faces As Integer) As Integer
		Return vertices - edges + faces
	End Function

	' Function to check if a space is simply connected
	Public Shared Function IsSimplyConnected(adjacencyList As List(Of Integer)()) As Boolean
		' Check if the space is connected
		If Not IsConnected(adjacencyList) Then
			Return False
		End If

		' Additional conditions for simple connectedness can be added based on the specific context
		Return True
	End Function

	' Function to check if a space is homotopy equivalent to a sphere
	Public Shared Function IsHomotopyEquivalentToSphere(adjacencyList As List(Of Integer)()) As Boolean
		' Additional conditions for homotopy equivalence can be added based on the specific context
		Return IsSimplyConnected(adjacencyList)
	End Function

	' Example usage
	'static void Main()
	'{
	' Example: Check if a graph is connected
	' List<int>[] graph = new List<int>[]
	'{
	'new List<int> {1, 2},
	'new List<int> {0, 2},
	'new List<int> {0, 1}
	'};

	'Console.WriteLine("Is the graph connected? " + IsConnected(graph));
	'}
End Class

File: Sigmath/Sigmath/imp/topology.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Sigmath
{
    using System;
    using System.Collections.Generic;
    using System.Linq;

    class Topology
    {
        // Function to check if a space is connected
        public static bool IsConnected(List<int>[] adjacencyList)
        {
            int numberOfVertices = adjacencyList.Length;
            bool[] visited = new bool[numberOfVertices];

            // Perform depth-first search starting from the first vertex
            DFS(adjacencyList, 0, visited);

            // Check if all vertices are visited
            return visited.All(v => v);
        }

        // Depth-first search to mark connected vertices
        private static void DFS(List<int>[] adjacencyList, int vertex, bool[] visited)
        {
            visited[vertex] = true;

            foreach (var neighbor in adjacencyList[vertex])
            {
                if (!visited[neighbor])
                {
                    DFS(adjacencyList, neighbor, visited);
                }
            }
        }

        // Function to check if a space is compact
        public static bool IsCompact(List<int>[] openCover)
        {
            foreach (var set in openCover)
            {
                if (!set.Any())
                {
                    return false; // An open cover must be non-empty
                }
            }

            // Additional conditions for compactness can be added based on the specific context
            return true;
        }

        // Function to calculate the Euler characteristic of a space
        public static int EulerCharacteristic(int vertices, int edges, int faces)
        {
            return vertices - edges + faces;
        }

        // Function to check if a space is simply connected
        public static bool IsSimplyConnected(List<int>[] adjacencyList)
        {
            // Check if the space is connected
            if (!IsConnected(adjacencyList))
            {
                return false;
            }

            // Additional conditions for simple connectedness can be added based on the specific context
            return true;
        }

        // Function to check if a space is homotopy equivalent to a sphere
        public static bool IsHomotopyEquivalentToSphere(List<int>[] adjacencyList)
        {
            // Additional conditions for homotopy equivalence can be added based on the specific context
            return IsSimplyConnected(adjacencyList);
        }

        // Example usage
        //static void Main()
        //{
            // Example: Check if a graph is connected
           // List<int>[] graph = new List<int>[]
            //{
            //new List<int> {1, 2},
            //new List<int> {0, 2},
            //new List<int> {0, 1}
            //};

            //Console.WriteLine("Is the graph connected? " + IsConnected(graph));
        //}
    }

}
